build/Dockerfile

```diff
@@ -12,38 +12,33 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM registry.access.redhat.com/ubi9/python-311 as wheel
 
-ARG WHEEL_VERSION=""
-USER root
-RUN --mount=type=cache,target=/root/.cache/pip \
-    python -m pip install --upgrade pip && \
-    python -m pip install build
-COPY tuning tuning
-COPY .git .git
-COPY pyproject.toml pyproject.toml
-# build wheel if wheel version is empty else download the wheel from PyPi
-RUN if [[ -z "${WHEEL_VERSION}" ]]; \
-    then python -m build --wheel --outdir /tmp; \
-    else pip download fms-hf-tuning==${WHEEL_VERSION} --dest /tmp --only-binary=:all: --no-deps; \
-    fi && \
-    ls /tmp/*.whl >/tmp/bdist_name
+## Global Args #################################################################
+ARG BASE_UBI_IMAGE_TAG=9.3-1552
+ARG PROTOC_VERSION=25.2
+ARG PYTORCH_INDEX="https://download.pytorch.org/whl"
 
+# match PyTorch version that was used to compile flash-attention v2 pre-built wheels
+# e.g. flash-attn v2.5.2 => torch ['1.12.1', '1.13.1', '2.0.1', '2.1.2', '2.2.0', '2.3.0.dev20240126']
+# https://github.com/Dao-AILab/flash-attention/blob/v2.5.2/.github/workflows/publish.yml#L47
+# use nightly build index for torch .dev pre-release versions
+ARG PYTORCH_VERSION=2.2.1
 
-FROM registry.access.redhat.com/ubi9/ubi AS release
+ARG PYTHON_VERSION=3.11
+ARG PYTHON_VERSION_SHORT=311
 
-ARG CUDA_VERSION=11.8.0
-ARG USER=tuning
-ARG USER_UID=1000
-ARG SET_NUM_PROCESSES_TO_NUM_GPUS=True
 
-USER root
+## Base Layer ##################################################################
+FROM registry.access.redhat.com/ubi9/ubi:${BASE_UBI_IMAGE_TAG} as base
+WORKDIR /app
+
+ARG PYTHON_VERSION
 
 RUN dnf update -y \
     && dnf remove -y --disableplugin=subscription-manager \
         subscription-manager \
         # we install newer version of requests via pip
-        python3.11-requests \
+    python${PYTHON_VERSION}-requests \
```

Comment:
dnf can not find the pthon3.11-requests package.
```
docker build --target base --progress=plain -t fms-hf-tuning:mytag . -f build/Dockerfile
...
No match for argument: python3.11-requests
```

Reply:
We could probably remove the installation of `make` and `procps` to further secure the final image since they are for development and debug.